### PR TITLE
gh-119400:  make_ssl_certs: update reference test data automatically, pass in expiration dates as parameters #119400 

### PR DESCRIPTION
**Pull request by kanavin**
As discussed here:
https://github.com/python/cpython/pull/107594

make_ssl_certs.py has a few shortcomings. In particular:

- it generates certificates, but does not update reference data in tests that use them, instead asking the user to copy paste the data by hand (expiration dates and serial numbers in particular)
- it is supposed to be run by hand and isn't executed in builds, which means its output has to be checked into git, cluttering the source tree, and complicating reviews of pull requests that change that output.
- expiration dates are hardcoded into the tool and can't be passed in as parameters

This pull request aims to address first and last issue, so then https://github.com/python/cpython/pull/107594 can move forward on top of them.

* Issue: gh-119400

### CODE REVIEW

**Comment by kanavin** [comment]
```
OSError: [Errno 30] Read-only file system: '/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/certdata'
```

This means CI is set up so that modifying the source tree is not possible. Suggestions? I still think it's worth making make_ssl_certs execution a part of the build, but the complication is that its output needs to be written into build dir, and both installation and tests needs to find it there.

**Comment by kanavin** [comment]
> ```
> OSError: [Errno 30] Read-only file system: '/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/certdata'
> ```
> 
> This means CI is set up so that modifying the source tree is not possible. Suggestions? I still think it's worth making make_ssl_certs execution a part of the build, but the complication is that its output needs to be written into build dir, and both installation and tests needs to find it there.

I've concluded that this is not feasible for now:
- needs invasive changes to Makefile
- requires openssl executable at build time
- breaks build reproducibility as every build is going to have different certificates installed, even if they're only used for testing.

I'll drop that from this PR, and make it only about not hardcoding reference certificate data and expiration parameters.

**Comment by kanavin** [comment]
This seems to be not getting any attention, is there something I can do to push it forward?

File: Lib/test/test_asyncio/utils.py [hunk 1]
```diff
@@ -56,24 +56,7 @@ def data_file(*filename):
...
-    'serialNumber': 'CB2D80995A69525C',
-    'subject': ((('countryName', 'XY'),),
-             (('localityName', 'Castle Anthrax'),),
-             (('organizationName', 'Python Software Foundation'),),
-             (('commonName', 'localhost'),)),
-    'subjectAltName': (('DNS', 'localhost'),),
-    'version': 3
-}
-
+PEERCERT = eval(open(data_file('certdata', 'keycert3.pem.reference')).read())
```
Comment (encukou) [inline Lib/test/test_asyncio/utils.py]: - the file is never closed
- `eval` is too powerful, let's use [`ast.literal_eval`](https://docs.python.org/3.8/library/ast.html#ast.literal_eval) to limit this data only.
```py
from ast import literal_eval
```

```suggestion
with open(data_file('certdata', 'keycert3.pem.reference')) as file:
    PEERCERT = literal_eval(file.read())
```

(same for the other `eval`)

File: Misc/NEWS.d/next/Build/2024-05-22-13-18-02.gh-issue-119400.WEt83v.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+make_ssl_certs: update reference test data automatically, do not hardcode expiration dates
```
Comment (encukou) [inline Misc/NEWS.d/next/Build/2024-05-22-13-18-02.gh-issue-119400.WEt83v.rst]:
```suggestion
``make_ssl_certs``, the script that prepares certificate data for the
test suite, now allows specifying expiration dates.
```

File: Lib/test/certdata/allsans.pem
Comment (encukou) [inline Lib/test/certdata/allsans.pem:1]: Could you omit the cert data changes from the PR?
There shouldn't be any meaningful change to these.

(The reference data would need to be regenerated to match the existing certs though.)

File: Lib/test/certdata/make_ssl_certs.py [hunk 3]
```diff
@@ -1,14 +1,18 @@
...
 import shutil
 import tempfile
 from subprocess import *
 
 startdate = "20180829142316Z"
-enddate = "20371028142316Z"
+enddate_default = "20371028142316Z"
+days_default = "7000"
+
+cmdlineargs = None
```
Comment (encukou) [inline Lib/test/certdata/make_ssl_certs.py]: Making `cmdlineargs` global means the functions can't work on their own.
Could you pass it (or its parts) into the functions that need it, as an argument?

**Review by encukou — COMMENTED** [review]
Sorry for the delay; my review queue grew too long and I had to trim it.
This looks great, but let's change a few details:

**Comment by kanavin** [comment]
@encukou Thanks, everything should be addressed now.

File: Lib/test/certdata/make_ssl_certs.py [hunk 4]
```diff
@@ -219,14 +223,22 @@ def make_ca():
...
 
 
-def print_cert(path):
+def write_cert_reference(path):
     import _ssl
-    pprint.pprint(_ssl._test_decode_cert(path))
+    refdata = pprint.pformat(_ssl._test_decode_cert(path))
+    print(refdata)
+    with open(path + '.reference', 'w') as f:
+        f.write(refdata)
```
Comment (encukou) [inline Lib/test/certdata/make_ssl_certs.py]: When I ran the script, the reference files were generated without trailing newlines. In this PR they have the newlines; I assume some tool added them as you added the files to Git.

Let's have the script print the newlines:
```suggestion
        print(refdata, file=f)
```

**Pull request merged**